test: rpc: add last block announcement time to getpeerinfo result

LarryRuane opened the pull request:
This PR adds `last_block_announcement` to the per-peer `getpeerinfo` RPC result. This is the most recent time that this peer was the first to notify our node of a new block (one that we didn't already know about), or zero if this peer has never been the first to notify us of a new block. This timestamp already exists internally and is used for stale-tip eviction logic; this PR exposes it at the RPC layer.

This PR started out as a suggestion for additional test coverage, see https://github.com/bitcoin/bitcoin/pull/26172#issuecomment-1259678260. It turned out that the easiest way to test (already-merged) #26172 is to add this field to `getpeerinfo` and have a functional test verify its value. But it may also be useful to have this result in its own right, similar to that RPC's existing `last_block` field -- it indicates something about the quality of our peers. It allows one to predict which peer will be evicted when the stale tip logic activates. (I'm not sure if that would be useful, but it may be.)

The functional test added here fails without #26172, which is the main goal.

This PR does not test the actual stale-tip eviction logic; that's difficult to do with a functional test. But it does test the correctness of the timestamp that the eviction logic depends on. #23352 is an attempt to test the eviction logic using a unit test, it's not ready to be merged yet. I think having both kinds of tests would be beneficial.

LarryRuane commented:
Force pushed to fix CI failure

satsie commented:
Concept ACK. I spent quite a bit of time trying to test #26172 (my findings are documented [here](https://github.com/satsie/bitcoin-notez/blob/master/bitcoin-core-development/testing-pr-26172.md)) and can confirm the difficulty of testing it with the current state of the test framework. 

I also support the suggestion to add `last_block_announcement` to `getpeerinfo` not only because it seems to be the easiest way to test 26172, but I agree it can be useful information to have when trying to learn more about one's peers, especially when this timestamp is already available internally.

MarcoFalke commented on src/net.h:
```diff
@@ -505,6 +505,10 @@ class CNode
 
+    /** UNIX epoch time of the last block announcement from this peer that we had
+     * not yet seen from any peer. Used for potential stale-tip eviction. */
+    std::atomic<std::chrono::seconds> m_last_block_announcement{0s};
```
If you change the type, why not use NodeSeconds (a time point)? While `std::chrono::seconds` (a duration) is used widely in the codebase, because all durations that denote a time point are since epoch, using a dedicated type such as `NodeSeconds` might be clearer.

dergoegge commented on src/net.h:
```diff
@@ -505,6 +506,10 @@ class CNode
 
+    /** UNIX epoch time of the last block announcement from this peer that we had
+     * not yet seen from any peer. Used for potential stale-tip eviction. */
+    std::atomic<std::chrono::seconds> m_last_block_announcement{0s};
```
I don't think this belongs in `CNode`. It is used in our outbound eviction logic which is part of net processing and it isn't used in net at all (imo, the eviction logic should be encapsulated in its own module entirely #25268 😉).

Also afaict, there is no need to move `m_last_block_announcement`, because you can just return it in `PeerManager::GetNodeStateStats`.

LarryRuane committed:
change m_last_block_announcement type from int64_t to NodeClock::time_point

This non-functional change makes it clear that this variable is a
time_point (point-in-time) rather than a duration. (An int64_t can be
either). This variable's use is modified in the following commits, so
we may as well modernize its type. This type is also mockable, which a
later test commit takes advantage of.

LarryRuane committed:
net: add m_last_block_announcement to CNodeStateStats

Copy its value from CNodeState. Unused until the next commit.

LarryRuane committed:
rpc: add last_block_announcement to the getpeerinfo output

This is the most recent time that this peer was the first to notify our
node of a new block (one that we didn't already know about), or zero if
this peer has never been the first to notify us of a new block. This can
be used to evaluate the quality (performance) of this peer, similar to
the existing `last_block` field.

This timestamp already exists internally and is used for stale-tip
eviction logic; this PR exposes it at the RPC layer. This will also be
used (in a later commit) to increase stale-tip test coverage.

LarryRuane committed:
test: add functional test for block announcement time tracking

Co-authored-by: naiyoma <lankas.aurelia@gmail.com>

LarryRuane committed:
doc: add release note for 27052

The `getpeerinfo` RPC now returns `last_block_announcement`.

LarryRuane commented:
Force pushed from becf38061b2625bcc629293fec0dce7c27292e14 to
- efc2e70f1763c6bf3f6b246202adb9d5f1f798ee [diff](https://github.com/bitcoin/bitcoin/compare/becf38061b2625bcc629293fec0dce7c27292e14..efc2e70f1763c6bf3f6b246202adb9d5f1f798ee) -- needed rebase
- 16d2903b8a9e6e980e24e04ab2d86c7e778f5764 [diff](https://github.com/bitcoin/bitcoin/compare/efc2e70f1763c6bf3f6b246202adb9d5f1f798ee..16d2903b8a9e6e980e24e04ab2d86c7e778f5764) -- make the suggested review changes
- 2e80e3bdcd9cac7878c0d14956ba9c57626da855 [diff](https://github.com/bitcoin/bitcoin/compare/16d2903b8a9e6e980e24e04ab2d86c7e778f5764..2e80e3bdcd9cac7878c0d14956ba9c57626da855) -- fix CI failure
- 6c564aafb3d5f61fffa0d6311ad688c47b14014a [diff](https://github.com/bitcoin/bitcoin/compare/2e80e3bdcd9cac7878c0d14956ba9c57626da855..6c564aafb3d5f61fffa0d6311ad688c47b14014a) -- add release notes, and small improvement to functional test

I'll take this out of draft now since I think it's ready for review.

kristapsk commented:
Concept ACK

LarryRuane commented:
Force-pushed to implement review suggestion to use `TicksSinceEpoch()` (thanks!)

dergoegge commented on src/net_processing.cpp:
```diff
@@ -5217,7 +5218,7 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
             // OUTBOUND_FULL_RELAY or MANUAL, protect it.
             if (!m_connman.MultipleManualOrFullOutboundConns(pnode->addr.GetNetwork())) return;
-            if (state->m_last_block_announcement < oldest_block_announcement || (state->m_last_block_announcement == oldest_block_announcement && pnode->GetId() > worst_peer)) {
+            if (!oldest_block_announcement.has_value() || (state->m_last_block_announcement == *oldest_block_announcement && pnode->GetId() > worst_peer)) {
```
This looks wrong to me, i think it should be:

```suggestion
            if (state->m_last_block_announcement < oldest_block_announcement.value_or(NodeSeconds::max()) || (state->m_last_block_announcement == *oldest_block_announcement && pnode->GetId() > worst_peer)) {
```

LarryRuane replied:
Thanks, good catch! I force-pushed to fix this in a slightly different way that seems more clear to me.

rkrux reviewed (approved):
ACK [b1c3826](https://github.com/bitcoin/bitcoin/pull/27052/commits/b1c3826ec093386c2ddec1d30ca3e777f8860892)

### Testing Methodology
1. Checked out PR
2. Built core
3. Ran `make check`
4. Ran `test/functional/test_runner.py`. 
5. Ran `bitcoind`
6. Ran `bcli getpeerinfo | jq '.[] | .last_block_announcement'` & got the following output that shows the presence of this new field `last_block_announcement` for nodes sending out new blocks to my node, 0 for the rest.
```
1710919881
0
1710919901
1710920179
0
0
0
0
0
0
```


Note: This `functional/feature_anchors.py` test did fail one time before it passed on the re-run, which I guess is due to unrelated error. PFB the error output.
```
stdout:
2024-03-20T07:07:27.561000Z TestFramework (INFO): PRNG seed is: 4633686694814387857
2024-03-20T07:07:27.562000Z TestFramework (INFO): Initializing test directory /var/folders/6v/mpspb4bx4zzf3xr2z96hgq9h0000gn/T/test_runner_₿_🏃_20240320_122543/feature_anchors_23
2024-03-20T07:07:28.673000Z TestFramework (INFO): When node starts, check if anchors.dat doesn't exist
2024-03-20T07:07:28.673000Z TestFramework (INFO): Add 2 block-relay-only connections to node
2024-03-20T07:07:28.942000Z TestFramework (INFO): Add 5 inbound connections to node
2024-03-20T07:07:29.682000Z TestFramework (INFO): Check node connections
2024-03-20T07:07:30.121000Z TestFramework (INFO): Check the addresses in anchors.dat
2024-03-20T07:07:30.122000Z TestFramework (INFO): Perturb anchors.dat to test it doesn't throw an error during initialization
2024-03-20T07:07:30.748000Z TestFramework (INFO): When node starts, check if anchors.dat doesn't exist anymore
2024-03-20T07:07:30.748000Z TestFramework (INFO): Ensure addrv2 support
2024-03-20T07:07:32.108000Z TestFramework (INFO): Add 256-bit-address block-relay-only connections to node
2024-03-20T07:07:32.761000Z TestFramework (INFO): Check for addrv2 addresses in anchors.dat
2024-03-20T07:07:32.845000Z TestFramework (INFO): Restarting node attempts to reconnect to anchors
2024-03-20T07:07:34.928000Z TestFramework (ERROR): Assertion failed

AssertionError: [node 0] Expected messages "['Trying to make an anchor connection to pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion:8333']" does not partially match log:
```

LarryRuane commented:
@rkrux - Thanks for the suggestions, force-pushed them.

naiyoma commented:
Concept ACK , Adding  `last_block_announcement ` field to the `getpeerinfo` RPC result is a valuable addition and also helps us get more coverage for untested `last_block_announcement`.

naiyoma commented on test/functional/p2p_block_times.py:
```diff
@@ -0,0 +1,83 @@
+        tip = int(node.getbestblockhash(), 16)
+        block = create_block(tip, create_coinbase(2))
+        block.solve()
+
```
nit: I think it would be useful to add an assertion to check the initial value of last_block_announcement
```suggestion
   self.log.info("Check that last_block_announcement is initially zero")
        peerinfo = node.getpeerinfo()[0]
        assert peerinfo['last_block_announcement'] == 0

```

naiyoma commented:
I tested to check if functional tests fail without the change from [https://github.com/bitcoin/bitcoin/pull/26172/](https://github.com/bitcoin/bitcoin/pull/26172/)

Reverted this commit [https://github.com/bitcoin/bitcoin/commit/9fcdb9f3a044330d3d7515fa35709102c98534d2](https://github.com/bitcoin/bitcoin/commit/9fcdb9f3a044330d3d7515fa35709102c98534d2)
make clean
./configure 
make 

The test fails as expected 

```
assert peerinfo['last_block_announcement'] == cur_time
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```
last_block_announcement doesn't get updated instead value is always default 0

LarryRuane commented:
Rebased for merge conflicts and to adopt @naiyoma's [review suggestion](https://github.com/bitcoin/bitcoin/pull/27052/files#r1736963290) (thank you!).

naiyoma commented on src/net_processing.cpp:
```diff
@@ -5189,7 +5189,7 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
         // to their network, counting both outbound-full-relay and manual peers.
         NodeId worst_peer = -1;
-        int64_t oldest_block_announcement = std::numeric_limits<int64_t>::max();
+        std::optional<NodeSeconds> oldest_block_announcement;
```
I'm not sure I understand why this is optional. Originally, it was set to a sentinel value.
I don't see a situation where it wouldn't have a value.
My suggestion → `NodeSeconds oldest_block_announcement = NodeSeconds::max();`
That way `.has_value() `checks and `*oldest_block_announcement` can be avoided.

naiyoma reviewed (commented):
Tested ACK cbe4603a90220f0ef0b21c4da68ee16791ad9034

Changes since last review: rebased and addressed nit fixes.

Also tested with `msg_cmpctblock`  - test passed as expected.

danielabrozzoni reviewed (commented):
Code Review d46a3a5cce4122c09a5b302dfec997e68ee9698d

I still haven't tested, I only looked at the code for now. The code looks ok, but there are some style improvements that I think should be picked up:

- I would like to see https://github.com/bitcoin/bitcoin/pull/27052#discussion_r2286284964, I think it would make the code more consistent and easier to understand
- In the first commit (e3ea5366cd5182525089cfd1a55dc675ff5faa46), it would be nice to have some more context in the commit message on why you're converting `m_last_block_announcement` to `NodeSeconds`. Something as easy as "preparatory for next commit" is fine, it's just to have some history in git as well 
- The PR description is good and gives a lot of insight as of why the PR is useful, but this context should be in the commit messages as well. I would add it to the commit message of the third commit (34b9122c2046f55901d3ccdcf9532f23e0c1326c)
- The first three commits should have the prefixes `rpc:` or `net:` in the commit message.

naiyoma reviewed (commented):
TestedACK d46a3a5cce4122c09a5b302dfec997e68ee9698d

Tested without the fix from PR #26172 by flipping the null check condition. 
The test fails as expected, catching `last_block_announcement` not being updated 
when a peer announces a new block.


```
assert peerinfo['last_block_announcement'] == cur_time
AssertionError
```


I have not been able to observe any  eviction logs, but I could see the last_block_announcement being set for my peers
Every 5.0s: bitcoin-cli  getpeerinfo | grep last_block_announcement                                                                                     ubuntu: Fri Mar 20 12:46:28 2026

    "last_block_announcement": 1773995600,
    "last_block_announcement": 0,
    "last_block_announcement": 0,
    "last_block_announcement": 1773999976,
    "last_block_announcement": 1773999593,
    "last_block_announcement": 0,
    "last_block_announcement": 1773997866,
    "last_block_announcement": 0,


micro nits: you can address this if you have to make another push

Consider adding a test case for when a peer sends a header that is new to us but has less chain work than our tip  `last_block_announcement`.

LarryRuane commented:
I think I've addressed (and accepted) all suggestions, thanks!

LarryRuane commented on test/functional/p2p_block_times.py:
```diff
@@ -0,0 +1,92 @@
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.disable_autoconnect = False
```
You're correct, the last of those 3 lines (`self.disable_autoconnect = False`) doesn't appear to be needed; I don't remember where I got it from (another test, probably). Removed, thanks.

sedited commented:
@LarryRuane can you fix the CI failures here?

naiyoma commented:
> I added this to the test, but could not get it to work (CI fails intermittently). I don't think it's very important, so I reverted it (force-pushed https://github.com/bitcoin/bitcoin/commit/136eed4a13c5c07c43675e27ddda14cb1a486661).

Yeah, this is not important. I just thought it would be a good addition to the test, but it shouldn’t block this PR from moving forward.
I did find the CI failure a bit odd. I’ll look into it. If I figure it out, I’ll share it here and likely add it as a follow-up.

luke-jr commented:
>In this very rare situation, if the timestamp gets updated (I don't think it will, but if it does), it's not much of a problem; it isn't critical that this timestamp is correct.

If it isn't important, then why expose it to RPC at all?

naiyoma commented:
> @naiyoma
> 
> > Consider adding a test case for when a peer sends a header that is new to us but has less chain work than our tip `last_block_announcement`.
> 
> I added this to the test, but could not get it to work (CI fails intermittently). I don't think it's very important, so I reverted it (force-pushed [136eed4](https://github.com/bitcoin/bitcoin/commit/136eed4a13c5c07c43675e27ddda14cb1a486661)). In this very rare situation, if the timestamp gets updated (I don't think it will, but if it does), it's not much of a problem; it isn't critical that this timestamp is correct. If you know how to add this to the test, please explain; otherwise, I think this is okay to merge.

I think I've been able to reproduce the CI failure.
The intermittent timeout depends on how fast the test runs.

When the test runs quickly, the first and second blocks are identical because
both are created within the same second and get the same timestamp. So
`blocks[-1].hash_hex == node.getbestblockhash()` is immediately True and
`send_blocks_and_test(success=True)` resolves instantly.

When there is a slight delay between the two create_block calls,
the blocks get different timestamps and are therefore different.
`block.hash_hex != block.hash_hex`, so `wait_until` keeps polling for
60 seconds waiting for the tip to change. This results in
a timeout.

A second issue is that the test was not correclty checking the edge case i suggested.
it was  passing because `received_new_header=0` 
we already had the block. The chainwork condition was never
tested.

My suggestion is to create a genuinely different block with an explicit timestamp:

    block2 = create_block(tip, create_coinbase(2), block.nTime + 1)
    block2.solve()
    headers_message2 = msg_headers()
    headers_message2.headers = [CBlockHeader(block2)]
    peer.send_and_ping(headers_message2)

This ensures `received_new_header=1` never seen this header 
but the condition still evaluates to false because `nChainWork` is equal not
greater:

    if (received_new_header && last_header.nChainWork > m_chainman.ActiveChain().Tip()->nChainWork)
    // True && (0006 > 0006) == False

I did generate mutants for this condition

```
if (received_new_header && last_header.nChainWork > m_chainman.ActiveChain().Tip()->nChainWork) {
        nodestate->m_last_block_announcement = Now<NodeSeconds>();
    }
```

```

<     if (received_new_header && last_header.nChainWork < m_chainman.ActiveChain().Tip()->nChainWork) {
---
>     if (received_new_header && last_header.nChainWork >= m_chainman.ActiveChain().Tip()->nChainWork) {
2899c2899
<     if (received_new_header && last_header.nChainWork <= m_chainman.ActiveChain().Tip()->nChainWork) {
---
>     if (received_new_header && last_header.nChainWork >= m_chainman.ActiveChain().Tip()->nChainWork) {
2899c2899
<     if (received_new_header || last_header.nChainWork > m_chainman.ActiveChain().Tip()->nChainWork) {
---
>     if (received_new_header && last_header.nChainWork >= m_chainman.ActiveChain().Tip()->nChainWork) {

```
Testing this edge case does kill these two surviving mutants (> -> >= and && -> ||). In my opinion, this is worth testing, so maybe it can be included here or in a follow-up.

sedited commented:
@LarryRuane do you plan to update your approach here with @naiyoma's suggestions?

LarryRuane commented:
@naiyoma - thanks for the review and mutation testing!

naiyoma reviewed (commented):
re-ACK e422fc50f39b5ddd0333b232a7437f3f3dadb73c

new test case added, test stylistic changes addressed, and commit messages prefixed, except for the first one, which could be net:

LarryRuane commented:
> ... except for the first one, which could be net:

Thanks for catching that, force-pushed to fix that (just the commit message, no other changes).

maflcko commented on src/test/denialofservice_tests.cpp:
```diff
@@ -200,7 +199,7 @@ BOOST_FIXTURE_TEST_CASE(stale_tip_peer_management, OutboundTest)
     // Update the last announced block time for the last
     // peer, and check that the next newest node gets evicted.
-    peerLogic->UpdateLastBlockAnnounceTime(vNodes.back()->GetId(), GetTime());
+    peerLogic->UpdateLastBlockAnnounceTime(vNodes.back()->GetId(), Now<NodeSeconds>());
```
nit in the first commit: This is not worse, but I don't think NodeSeconds is the right type here.

It seems more consistent to use the native clock type (and resolution), that is NodeClock::time_point.

I've pushed a commit that does this to fa67c651510db5e571fa82fe692ac33e48c451a4 (https://github.com/bitcoin/bitcoin/pull/35315/commits)

sedited commented:
@LarryRuane do you want to address maflcko's nit?

LarryRuane commented:
CI failures should be fixed by rebasing (there was a hidden merge conflict); functional test framework's `create_block()`'s call signature changed to require the block time to be given as a keyword argument, rather than in position 3. The fix is to add the `ntime=` keyword to the one place that the new functional test, `p2p_block_times.py`, calls that function.
